src/Network/GRPC/MQTT/Core.hs

```diff
@@ -28,54 +29,92 @@ import Data.Conduit.Network.TLS (
-setCallback cb (Secured cfg) = Secured cfg{_msgCB = cb}
-setCallback cb (Unsecured cfg) = Unsecured cfg{_msgCB = cb}
+defaultMGConfig :: MQTTGRPCConfig
+defaultMGConfig =
+  MQTTGRPCConfig
+    { useTLS = False
+    , mqttMsgSizeLimit = 128000
+    , _cleanSession = True
+    , _lwt = Nothing
+    , _msgCB = NoCallback
+    , _protocol = Protocol311
+    , _connProps = []
+    , _hostname = ""
+    , _port = 1883
+    , _connID = []
+    , _username = Nothing
+    , _password = Nothing
+    , _connectTimeout = 180000000
+    , _tlsSettings = TLSSettingsSimple False False False
+    }
 
--- | Utility for setting the connection ID within the 'MQTTConfig'
-setConnectionId :: String -> MQTTConnectionConfig -> MQTTConnectionConfig
-setConnectionId cid (Secured cfg) = Secured cfg{_connID = cid}
-setConnectionId cid (Unsecured cfg) = Unsecured cfg{_connID = cid}
+-- | Project 'MQTTConfig' from 'MQTTGRPCConfig'
+getMQTTConfig :: MQTTGRPCConfig -> MQTTConfig
+getMQTTConfig MQTTGRPCConfig{..} = MQTTConfig{..}
 
 -- | Connect to an MQTT broker
-connectMQTT :: MonadIO m => MQTTConnectionConfig -> m MQTTClient
-connectMQTT = \case
-  Secured cfg@MQTTConfig{..} -> liftIO $ runMQTTConduit runTLS cfg
-   where
-    runTLS :: (MQTTConduit -> IO ()) -> IO ()
-    runTLS f = runTLSClient tlsCfg (f . toMQTTConduit)
+connectMQTT :: MonadIO m => MQTTGRPCConfig -> m MQTTClient
+connectMQTT cfg@MQTTGRPCConfig{..} = liftIO $ do
+  let runner = if useTLS then runTLS else runTCP
+  runMQTTConduit runner (getMQTTConfig cfg)
+ where
+  runTLS :: (MQTTConduit -> IO ()) -> IO ()
+  runTLS f = runTLSClient tlsCfg (f . toMQTTConduit)
 
-    tlsCfg :: TLSClientConfig
-    tlsCfg = (tlsClientConfig _port (encodeUtf8 _hostname)){tlsClientTLSSettings = _tlsSettings}
-  Unsecured cfg@MQTTConfig{..} -> liftIO $ runMQTTConduit runTCP cfg
-   where
-    runTCP :: (MQTTConduit -> IO ()) -> IO ()
-    runTCP f = runTCPClient tcpCfg (f . toMQTTConduit)
+  tlsCfg :: TLSClientConfig
+  tlsCfg = (tlsClientConfig _port (encodeUtf8 _hostname)){tlsClientTLSSettings = _tlsSettings}
 
-    tcpCfg :: ClientSettings
-    tcpCfg = clientSettings _port (encodeUtf8 _hostname)
+  runTCP :: (MQTTConduit -> IO ()) -> IO ()
+  runTCP f = runTCPClient tcpCfg (f . toMQTTConduit)
 
-toMQTTConduit :: AppData -> MQTTConduit
-toMQTTConduit ad = (appSource ad, appSink ad)
+  tcpCfg :: ClientSettings
+  tcpCfg = clientSettings _port (encodeUtf8 _hostname)
+
+  toMQTTConduit :: AppData -> MQTTConduit
+  toMQTTConduit ad = (appSource ad, appSink ad)
 
 -- | Period for heartbeat messages
 heartbeatPeriodSeconds :: NominalDiffTime
-heartbeatPeriodSeconds = 10
\ No newline at end of file
+heartbeatPeriodSeconds = 10
+
+{- | Topics are strictly a subset of 'Filter's so this conversion is always safe
+ | This is a bit of a hack, but there is a upstream PR to add this to the net-mqtt
+ | library: https://github.com/dustin/mqtt-hs/pull/22
+-}
+toFilter :: Topic -> Filter
+toFilter = fromJust . mkFilter . unTopic
```

Comment:
Can you `coerce` instead?

```suggestion
toFilter = coerce
```

Reply:
Unfortunately, no. The library does not export the constructors. I think `unsafeCoerce` would work, but then we're right back in unsafe territory. 
Fortunately, my `net-mqtt` upstream PR was already accepted and merged in. So, I'll be able to remove this and just import `toFilter` from there.